core/qthelper.cpp

```diff
@@ -1223,23 +1240,17 @@ void learnImages(const QDir dir, int max_recursions)
 
 extern "C" const char *local_file_path(struct picture *picture)
 {
-	QString hashString = picture->hash;
-	if (hashString.isEmpty()) {
-		QByteArray hash = hashFile(picture->filename);
-		free(picture->hash);
-		picture->hash = strdup(hash.toHex().data());
-	}
-	QString localFileName = fileFromHash(picture->hash);
-	if (localFileName.isEmpty())
-		localFileName = picture->filename;
-	return strdup(qPrintable(localFileName));
+	return strdup(qPrintable(localFilePath(picture->filename)));
 }
 
 extern "C" bool picture_exists(struct picture *picture)
 {
-	QString localFilename = fileFromHash(picture->hash);
-	QByteArray hash = hashFile(localFilename);
-	return same_string(hash.toHex().data(), picture->hash);
+	QString localPath = localFilePath(picture->filename);
+	if (localPath.isEmpty())
+		return false;
+	QByteArray hash = hashFile(localPath);
+	QMutexLocker locker(&hashOfMutex);
```

Comment:
Why do you lock here? You don't access the the QHashes here. In fact, getHash() hangs because it cannot obtain the lock itself. Removing this lock seems to make it work for me.

Reply:
That was an accident.